### ISSUE TITLE
Support for Java Simple Plugin System (with Jar files)

### ISSUE THREAD

**Issue by farko88** [spring-projects/spring-boot#1546]
Hi, 

it will be good if there is some plugin system for Spring Boot. As many Spring Boot applications is distributed as jar files so it will be core of system, when needed additional features or want to remove some feature just add/remove jar from some defined folder ex. "plugins". I recently tried to achieve this but without success i can't get controller and other annotations provided by Spring working from plugin jar. Tried with JSPF (Java Simple Plugin Framework), getting plugin injected but can't get @Controller inside the plugin jar working.

If anyone can help me with some details i am ready to work on this feature.

**Comment by tvollstaedt** [comment]
@farko88 Can you please re-upload your test repo, I'm currently struggling with exact the same issue. Loaded the JAR via PluginManager, but Spring won't recognize any of the beans.

**Comment by laxika** [comment]
This would be really cool to have. I do need a plugin framework for spring both at work and at my personal projects.

**Comment by philwebb** [comment]
I think the current system of using `spring.factories` is sufficient. We're not keen to add anything more complex than that.

**Comment by Jav1st** [comment]
Hi guys,

I tried your spring.factories and it works greate.
But this aproach has one sufficient drawback.
To extend Spring Boot application is required that each plugin has to be available into classpath at load-time.
So classpath will be shared among all plugins.
For simple cases it is not a problem. For more complex cases it might lead to jar hell issue 
(if several plugins have different versions of the same dependencies, not provided by the main application, you might have some weird effects)
To solve it each plugin should be loaded in a separate class loader.
After that spring.factories should be registered into apllication context and latter should be refreshed.

Look at my desirable code snippet below:
 === main_app.jar ===
         
     interface SpringBootPlugin{

      default void init(ConfigurableApplicationContext applicationContext){
         applicationContext.loadFactories(pluginClassLoader()); 
      }

      void start();
      void destroy();
      ClassLoader pluginClassLoader();
    }



    @SpringBootApplication
    public class MainApplication{

       public static void main(String[] args) {
	    ConfigurableApplicationContext configurableApplicationContext = SpringApplication.run(MainApplication.class, args); 
        List<SpringBootPlugin> plugins = PluginLoader.loadPlugins("./plugindir"); // use reflection API to load each jar file in a separate class loader
        
        for (SpringBootPlugin plugin:plugins){
          plugin.init(configurableApplicationContext);
        }         

        configurableApplicationContext.refresh(); // call refresh again to wire/activate run-time factories. I know that now it is not possible to call it more than once

        for (SpringBootPlugin plugin:plugins){
          plugin.start();
        }   

      }

    }

=== custom_plugin.jar ===

    META-INF/spring.factories

    public class customPlugin implements SpringBootPlugin{
     // some implementation of the start and destroy methods
    }

    @Configuration
    @ComponentScan(basePackages = "com.plugins.custom")
    public class PluginAutoConfiguration{
      // plugin beans declarations
    }

So, as you can see, I need a explicit feature to repeat loading auto configurations from spring.factories.
So I am wondering is there any way to achieve that?

**Comment by dsyer** [comment]
I don't think it's at all clear what you want to achieve here, and the proposed implementation is a bit fuzzy (to say the least). If you have a crisper description of what you need, and maybe an example use case, it might be worth dropping that in, and we can try and understand.

**Comment by Jav1st** [comment]
What exactly is not clear here?

I want to build a "real" pluggable apllication based on Spring Boot. When I say "real" I mean that each plugin should be isolated in its own class loader.
As you said before you've already provided a very simple (and successful) "plugin" mechanism based on @AutoConfiguration and spring.factories.
I've tried it and it looks like very nice. But now this mechanism works on start time only (when application context is constructing). 
All configurations which are found in jars(via spring.factories) from the shared classpath then loaded into context. 
After that (when context is already constructed) it seems no way now to load additional dynamic factories and refresh context.

I just need a possibility to programatically load auto configurations from each plugin's spring.factories into already constructed apllication context. 
This is what I do in supposed plugin's init method at code snippet above.

**Comment by dsyer** [comment]
> I just need a possibility to programatically load auto configurations from each plugin's spring.factories into already constructed application context.

I don't see how that could work without a lot more detailed specification, or maybe some prototype code if you have some you would like to share. A class loader is quite a hard boundary. How would the beans from the plugin communicate with the ones from the other class loaders? You probably need some sort of service registry and a way to keep track of classes or interfaces that are shared between all contexts. There's a lot of plumbing to do.

In short, I don't think we are going to get into that game at all, but for sure if we do it will take a lot of code, and some more concrete proposals (not just a snippet).

**Comment by martin-g** [comment]
@Jav1st Check https://github.com/kamranzafar/JCL. I haven't used it myself but its README claims that it supports what you ask for.

**Comment by philwebb** [comment]
> I want to build a "real" pluggable Spring Boot application. When I say "real" I mean that each plugin should be isolated in its own class loader.

This sounds like a great deal of work and it's just not something we're keen to invest in. Our limited resources need to be invested in the areas that provide the most benefit to all our users.
